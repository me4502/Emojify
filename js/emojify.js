const translate = require('moji-translate');

export function emojify() {
    const textArea = document.querySelector("#text_input");
    const inputText = textArea.value;
    const lines = inputText.split(/\n/);
    const outputLines = [];
    const missingWords = [];
    for (const line of lines) {
        const words = line.split(" ");
        let lastTriple = 1000;
        for (const i in words) {
            const word = words[i];
            const cleanedWord = cleanWord(word);
            let emoji = emojifyWord(cleanedWord);
            if (emoji) {
                if (Math.random() > 0.95 && lastTriple > 3) {
                    emoji = emoji + emoji + emoji;
                    lastTriple = 0;
                } else {
                    lastTriple++;
                }
                words[i] = word + " " + emoji + " ";
            } else {
                missingWords.push(cleanedWord);
            }
        }
        outputLines.push(words.join(" "));
    }
    const textOutput = document.querySelector("#text_output");
    textOutput.value = outputLines.join("\n");
    textOutput.hidden = false;
    M.textareaAutoResize(textOutput);

    console.log("The following words were missing emojis :( " + missingWords.join(","));
}

function cleanWord(word) {
    word = word.trim()
        .replace("?", "")
        .replace("!", "")
        .replace(",", "")
        .replace("\"", "")
        .replace(".", "");
    return word.toLowerCase();
}

function emojifyWord(word) {
    let emoji = translate.getEmojiForWord(word);
    if (!emoji) {
        // Do some custom overrides
        if (word.endsWith("'s") && translate.getEmojiForWord(word.substring(0, word.length - 2))) {
            return translate.getEmojiForWord(word.substring(0, word.length - 2));
        }
        if (word.endsWith("'") && translate.getEmojiForWord(word.substring(0, word.length - 1) + "g")) {
            return translate.getEmojiForWord(word.substring(0, word.length - 1) + "g");
        }

        // Specific words
        if (word === 'yell' || word === 'yelling') {
            return "ðŸ¤¬";
        }
        if (word === 'tell') {
            return "ðŸ—£";
        }
    }
    if (word === emoji) {
        return undefined;
    }
    return emoji;
}