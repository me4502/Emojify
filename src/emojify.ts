import translate from 'moji-translate';
import { lib } from 'emojilib';
import * as emojiOverridesJson from './overrides.json';

type EmojiMap = { [key: string]: string[] };

const emojiOverrides = emojiOverridesJson as EmojiMap;

let _emojiMap: EmojiMap | undefined = undefined;

function getEmojiMap(): EmojiMap {
    if (_emojiMap) {
        return _emojiMap;
    }
    _emojiMap = Object.keys(lib).reduce((obj, emojiKey) => {
        const emoji = lib[emojiKey];
        obj[emoji.char] = emoji.keywords.reduce((a, b) => {
            a.push(...b.split('-'));
            return a;
        }, [] as string[]);
        obj[emoji.char].push(emojiKey);
        obj[emoji.char].push(...emojiKey.split('_'));
        obj[emoji.char] = obj[emoji.char].filter(
            w =>
                w !== 'and' &&
                w !== 'in' &&
                w !== 'with' &&
                w !== 'of' &&
                w !== 'a'
        );
        return obj;
    }, {} as EmojiMap);

    for (const override in emojiOverrides) {
        _emojiMap[override] = (_emojiMap[override] ?? []).concat(
            emojiOverrides[override]
        );
    }

    return _emojiMap!;
}

interface EmojifyOptions {
    replace?: boolean;
    /* Chance of replacing words with emojis when replace is enabled. 1 means always 0 means never. */
    replaceChance?: number;
    tripleCooldown?: number;
    tripleChance?: number;
    multimojiChance?: number;
}

export function emojify(
    inputText: string,
    {
        replace = false,
        replaceChance = 1,
        tripleCooldown = 3,
        tripleChance = 0.05,
        multimojiChance = 0.1
    }: EmojifyOptions = {}
): string {
    const lines = inputText.split(/\n/);
    const outputLines = [];
    const missingWords = [];

    for (const line of lines) {
        const words = line.split(' ');
        let lastTriple = 99999;

        for (const i in words) {
            const word = words[i];
            const cleanedWord = cleanWord(word);
            let emojis = findEmojisForWord(cleanedWord);

            if (emojis.length > 0) {
                function grabEmoji(): string {
                    const emojiIndex = Math.floor(
                        Math.random() * emojis.length
                    );
                    const emoji = emojis[emojiIndex];
                    emojis.splice(emojiIndex, 1);
                    return emoji;
                }

                const emoji = grabEmoji();

                let insert = emoji;

                while (Math.random() < multimojiChance && emojis.length > 1) {
                    insert += grabEmoji();
                }

                if (
                    Math.random() < tripleChance &&
                    lastTriple > tripleCooldown
                ) {
                    insert = insert + insert + insert;
                    lastTriple = 0;
                } else {
                    lastTriple++;
                }

                if (replace && Math.random() < replaceChance) {
                    words[i] = insert;
                } else {
                    words[i] = word + ' ' + insert + ' ';
                }
            } else {
                missingWords.push(cleanedWord);
            }
        }
        outputLines.push(words.join(' '));
    }
    return outputLines.join('\n');
}

function cleanWord(word: string): string {
    word = word
        .trim()
        .replace('?', '')
        .replace('!', '')
        .replace(',', '')
        .replace(':', '')
        .replace('"', '')
        .replace('"', '')
        .replace('.', '');
    return word.toLowerCase();
}

function findEmojisForWord0(word: string): string[] {
    const foundEmojis: string[] = [];

    if (word === 'my') {
        // Easy hack fix for a common flag issue.
        word = 'i';
    }

    // Maintain a whitelist of <3 letter words
    if (
        word.length <= 2 &&
        word !== 'i' &&
        word !== 'he' &&
        word !== 'we' &&
        word !== 'no'
    ) {
        return foundEmojis;
    }

    if (translate.isMaybeAlreadyAnEmoji(word)) {
        return foundEmojis;
    }

    if (emojiOverrides[word]) {
        foundEmojis.push(...emojiOverrides[word]);
    }

    const emojiMap = getEmojiMap();

    for (let emoji in emojiMap) {
        if (emojiMap[emoji].includes(word)) {
            foundEmojis.push(emoji);
        }
    }

    return foundEmojis;
}

const emojiModFunctions: ((word: string) => string | undefined)[] = [
    w => w,
    w => (w.length > 1 ? `${w}s` : undefined),
    w => (w.length > 1 ? `${w}ing` : undefined),
    w => (w.length > 1 ? `${w}ed` : undefined),
    w => (w.endsWith("'s") ? w.substring(0, w.length - 2) : undefined),
    w => (w.endsWith("'") ? `${w.substring(0, w.length - 1)}g` : undefined),
    w =>
        w.length > 2 && w.endsWith('s')
            ? w.substring(0, w.length - 1)
            : undefined,
    w => (w.endsWith('ing') ? w.substring(0, w.length - 3) : undefined),
    w => (w.endsWith('ing') ? `${w.substring(0, w.length - 3)}e` : undefined),
    w =>
        w.length > 5 && w.endsWith('ing')
            ? w.substring(0, w.length - 4)
            : undefined,
    w => (w.endsWith('er') ? w.substring(0, w.length - 2) : undefined),
    w => (w.endsWith('er') ? `${w.substring(0, w.length - 2)}e` : undefined),
    w =>
        w.length > 4 && w.endsWith('er')
            ? w.substring(0, w.length - 3)
            : undefined
];

function findEmojisForWord(word: string): string[] {
    const foundEmojis: string[] = [];

    // Try the word, and various permutations.
    word = word.trim().toLowerCase();

    // Split hyphenated words
    if (word.includes('-')) {
        word.split('-')
            .map(findEmojisForWord)
            .forEach(words => foundEmojis.push(...words));
    }

    // Run it for each functor
    emojiModFunctions
        .map(func => func(word))
        .filter(w => w)
        .forEach(w => foundEmojis.push(...findEmojisForWord0(w!)));

    return foundEmojis;
}
