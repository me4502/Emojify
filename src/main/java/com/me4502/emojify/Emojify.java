package com.me4502.emojify;

import com.me4502.emojify.app.EmojiApp;

import java.io.IOException;

public class Emojify {

    public static void main(String[] args) {
        try {
            EmojiApp.getInstance().load();
        } catch (IOException e) {
            // If an exception makes it here, runtime.
            throw new RuntimeException(e);
        }
    }
}
