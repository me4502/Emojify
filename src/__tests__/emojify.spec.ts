import { emojify } from '../emojify';

function mockRandom(value: number, fn: Function): (...args: any[]) => any {
    return function (this: unknown, ...args: any[]) {
        jest.spyOn(global.Math, 'random').mockReturnValue(value);
        const returnValue = fn.apply(this, args);
        jest.spyOn(global.Math, 'random').mockRestore();
        return returnValue;
    };
}

const SIMPLE_TEXT = 'he see carrot tree house';

describe('emojify', () => {
    it(
        'should emojify some simple text',
        mockRandom(0.0, () => {
            expect(
                emojify(SIMPLE_TEXT, {
                    multimojiChance: 0,
                    replace: false,
                    replaceChance: 0,
                    tripleChance: 0,
                    tripleCooldown: 0
                })
            ).toMatchSnapshot();
        })
    );

    it(
        'should emojify some simple text with multimoji',
        mockRandom(0.0, () => {
            expect(
                emojify(SIMPLE_TEXT, {
                    multimojiChance: 1.0,
                    replace: false,
                    replaceChance: 0,
                    tripleChance: 0,
                    tripleCooldown: 0
                })
            ).toMatchSnapshot();
        })
    );

    it(
        'should emojify some simple text with replace mode',
        mockRandom(0.0, () => {
            expect(
                emojify(SIMPLE_TEXT, {
                    multimojiChance: 0,
                    replace: true,
                    replaceChance: 1.0,
                    tripleChance: 0,
                    tripleCooldown: 0
                })
            ).toMatchSnapshot();
        })
    );
});
