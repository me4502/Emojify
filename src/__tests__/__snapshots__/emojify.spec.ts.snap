// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emojify should emojify some simple text 1`] = `"he 💁‍♂️  see 🙈  carrot 🥕  tree 🌲  house 🪴  is nice relieve 😮‍💨  relief 😅 "`;

exports[`emojify should emojify some simple text with multimoji 1`] = `"he 💁‍♂️  see 🙈👀  carrot 🥕  tree 🌲🌳🌴🍃🎄  house 🪴🛖🏚️🏠🏡🚪  is nice relieve 😮‍💨😅  relief 😅😮‍💨 "`;

exports[`emojify should emojify some simple text with replace mode 1`] = `"💁‍♂️ 🙈 🥕 🌲 🪴 is nice 😮‍💨 😅"`;

exports[`emojify should emojify some simple text with triple mode 1`] = `"he 💁‍♂️💁‍♂️💁‍♂️  see 🙈🙈🙈  carrot 🥕🥕🥕  tree 🌲🌲🌲  house 🪴🪴🪴  is nice relieve 😮‍💨😮‍💨😮‍💨  relief 😅😅😅 "`;
